### PR TITLE
Pv2 2302 3 performance improvements

### PR DESCRIPTION
**Pull request by tom-gough**

### CODE REVIEW

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs [hunk 1]
```diff
@@ -11,68 +13,92 @@ namespace SFA.DAS.Provider.Events.Infrastructure.Data
...
+
+                var count = await GetTransferCount(senderAccountId, receiverAccountId, academicYear, collectionPeriod, connection);
+
+                var pagedResults = PageResults(result.ToList(), page, pageSize, count);
+
+                return pagedResults;
+            }
+        }
+
+        private async Task<IEnumerable<TransferEntity>> GetUnPaginatedTransfers(int page, int pageSize, long? senderAccountId, long? receiverAccountId, int? academicYear, int? collectionPeriod, SqlConnection connection)
```
Comment (carlbm) [inline src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs:52]: Rename to GetPaginatedTransfers

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs [hunk 2]
```diff
@@ -11,68 +13,92 @@ namespace SFA.DAS.Provider.Events.Infrastructure.Data
...
+            AddPagingQueryParameters(page, pageSize, sqlBuilder);
+
+            var result = await connection.QueryAsync<TransferEntity>(query.RawSql, query.Parameters).ConfigureAwait(false);
+
+            return result;
+        }
+
+        private static async Task<int> GetTransferCount(long? senderAccountId, long? receiverAccountId, int? academicYear, int? collectionPeriod, SqlConnection connection)
+        {
+            var countSQlBuilder = new SqlBuilder();
```
Comment (carlbm) [inline src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs:68]: Lowercase Q

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs [hunk 3]
```diff
@@ -11,68 +13,92 @@ namespace SFA.DAS.Provider.Events.Infrastructure.Data
...
+            var countQuery = countSQlBuilder.AddTemplate(CountSqlTemplate);
+
+            BuildQueryParameters(senderAccountId, receiverAccountId, academicYear, collectionPeriod, countSQlBuilder);
+
+            var count = await connection.QueryFirstOrDefaultAsync<int>(
```
Comment (carlbm) [inline src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs:73]: You can use `ExecuteScalarAsync` to get the single returned value

**Review by carlbm — CHANGES_REQUESTED** [review]

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsPaymentRepository.cs [hunk 4]
```diff
@@ -46,70 +43,86 @@ public class DcfsPaymentRepository : DcfsRepository, IPaymentRepository
...
+		        query.RawSql,
+		        query.Parameters
+	        ).ConfigureAwait(false);
+	        
+	        return result;
+        }
+
+        private static async Task<int> GetPaymentCount(string employerAccountId, int? academicYear, int? collectionPeriod, long? ukprn, SqlConnection connection)
+        {
+	        var countSQlBuilder = new SqlBuilder();
```
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsPaymentRepository.cs:95]: fixed

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs (hunk 2 above)
Comment (tom-gough) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs:68]: fixed

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsPaymentRepository.cs [hunk 5]
```diff
@@ -46,70 +43,86 @@ public class DcfsPaymentRepository : DcfsRepository, IPaymentRepository
...
+
+	            var pagedResults = PageResults(unPaginatedPayments.ToList(), page, pageSize, count);
 
-                var pagedResults = PageResults(result.ToList(), page, pageSize);
                 return pagedResults;
             }
         }
 
-        public async Task<PaymentStatistics> GetStatistics()
+        private static async Task<IEnumerable<PaymentEntity>> GetUnPaginatedPayments(int page, int pageSize, string employerAccountId, int? academicYear, int? collectionPeriod, long? ukprn, SqlConnection connection)
```
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsPaymentRepository.cs:76]: fixed, I think this was just a result of a rename followed by a refactor

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs (hunk 1 above)
Comment (tom-gough) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs:52]: fixed

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsPaymentRepository.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsPaymentRepository.cs:76]: I have also updated PageResults to be AddPagingInformation which is more descriptive of what is actually happening now

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs (hunk 3 above)
Comment (tom-gough) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsTransferRepository.cs:73]: updated here and in payments repository

**Review by carlbm — APPROVED** [review]

**Review by carlbm — APPROVED** [review]

**Pull request merged**